File: packages/vmcp-dsl/src/patterns/map-each.ts
/**
 * Map-each pattern builder
 */

import type {
  PatternSpec,
  MapEachSpec,
} from '../types.js';

/**
 * Builder for map-each patterns
 */
export class MapEachBuilder {
  private inner?: { tool: string } | { pattern: PatternSpec };

  /**
   * Apply a tool to each element
   */
  tool(name: string): this {
    this.inner = { tool: name };
    return this;
  }

  /**
   * Apply a pattern to each element
   */
  pattern(spec: PatternSpec): this {
    this.inner = { pattern: spec };
    return this;
  }

  /**
   * Build the map-each pattern spec
   */
  build(): PatternSpec {
    if (!this.inner) {
      throw new Error('Inner operation (tool or pattern) is required');
    }
    return { mapEach: { inner: this.inner } };
  }

  /**
   * Get the raw spec
   */
  spec(): MapEachSpec {
    if (!this.inner) {
      throw new Error('Inner operation (tool or pattern) is required');
    }
    return { inner: this.inner };
  }
}

/**
 * Create a map-each pattern
 *
 * @example
 * // Apply a tool to each element
 * const fetchAll = mapEach()
 *   .tool('fetch_document')
 *   .build();
 *
 * // Apply a pattern to each element
 * const normalizeAll = mapEach()
 *   .pattern(schemaMap().field('title', '$.name').build())
 *   .build();
 */
export function mapEach(): MapEachBuilder {
  return new MapEachBuilder();
}

/**
 * Shorthand for creating a map-each with a tool
 */
export function mapEachTool(toolName: string): PatternSpec {
  return { mapEach: { inner: { tool: toolName } } };
}

/**
 * Shorthand for creating a map-each with a pattern
 */
export function mapEachPattern(spec: PatternSpec): PatternSpec {
  return { mapEach: { inner: { pattern: spec } } };
}

File: crates/agentgateway/src/mcp/registry/patterns/map_each.rs
// Map Each pattern types

use serde::{Deserialize, Serialize};

use super::PatternSpec;

/// MapEachSpec applies an operation to each element of an array
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MapEachSpec {
	/// The operation to apply to each element
	pub inner: MapEachInner,
}

impl MapEachSpec {
	/// Create a MapEach that calls a tool for each element
	pub fn tool(name: impl Into<String>) -> Self {
		Self { inner: MapEachInner::Tool(name.into()) }
	}

	/// Create a MapEach that applies a pattern for each element
	pub fn pattern(spec: PatternSpec) -> Self {
		Self { inner: MapEachInner::Pattern(Box::new(spec)) }
	}

	/// Get the names of tools referenced by this map-each
	pub fn referenced_tools(&self) -> Vec<&str> {
		self.inner.referenced_tools()
	}
}

/// The inner operation of a MapEach
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum MapEachInner {
	/// Tool name to call for each element
	Tool(String),

	/// Pattern to apply for each element
	Pattern(Box<PatternSpec>),
}

impl MapEachInner {
	/// Get the names of tools referenced by this inner operation
	pub fn referenced_tools(&self) -> Vec<&str> {
		match self {
			MapEachInner::Tool(name) => vec![name.as_str()],
			MapEachInner::Pattern(p) => p.referenced_tools(),
		}
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_parse_map_each_tool() {
		let json = r#"{
			"inner": { "tool": "fetch_document" }
		}"#;

		let map_each: MapEachSpec = serde_json::from_str(json).unwrap();
		assert!(matches!(map_each.inner, MapEachInner::Tool(ref name) if name == "fetch_document"));
	}

	#[test]
	fn test_parse_map_each_pattern() {
		let json = r#"{
			"inner": {
				"pattern": {
					"schemaMap": {
						"mappings": {
							"title": { "path": "$.name" }
						}
					}
				}
			}
		}"#;

		let map_each: MapEachSpec = serde_json::from_str(json).unwrap();
		assert!(matches!(map_each.inner, MapEachInner::Pattern(_)));
	}

	#[test]
	fn test_builder_tool() {
		let map_each = MapEachSpec::tool("my_tool");
		assert!(matches!(map_each.inner, MapEachInner::Tool(ref name) if name == "my_tool"));
	}

	#[test]
	fn test_referenced_tools() {
		let map_each = MapEachSpec::tool("fetch");
		let refs = map_each.referenced_tools();
		assert_eq!(refs, vec!["fetch"]);
	}
}

